src/app/clusters/general-commissioning-server/general-commissioning-server.cpp

```diff
@@ -218,60 +373,90 @@ bool emberAfGeneralCommissioningClusterCommissioningCompleteCallback(
     auto & failSafe               = Server::GetInstance().GetFailSafeContext();
     auto & fabricTable            = Server::GetInstance().GetFabricTable();
 
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
     ChipLogProgress(FailSafe, "GeneralCommissioning: Received CommissioningComplete");
 
     Commands::CommissioningCompleteResponse::Type response;
+
+    // Fail-safe must be armed
     if (!failSafe.IsFailSafeArmed())
     {
         response.errorCode = CommissioningErrorEnum::kNoFailSafe;
+        commandObj->AddResponse(commandPath, response);
+        return true;
     }
-    else
+
+    SessionHandle handle = commandObj->GetExchangeContext()->GetSessionHandle();
+
+    // Ensure it's a valid CASE session
+    if ((handle->GetSessionType() != Session::SessionType::kSecure) ||
+        (handle->AsSecureSession()->GetSecureSessionType() != SecureSession::Type::kCASE) ||
+        (!failSafe.MatchesFabricIndex(commandObj->GetAccessingFabricIndex())))
+    {
+        response.errorCode = CommissioningErrorEnum::kInvalidAuthentication;
+        ChipLogError(FailSafe, "GeneralCommissioning: Got commissioning complete in invalid security context");
+        commandObj->AddResponse(commandPath, response);
+        return true;
+    }
+
+#if CHIP_CONFIG_TC_REQUIRED
+    TermsAndConditionsProvider * const termsAndConditionsProvider = Server::GetInstance().GetTermsAndConditionsProvider();
+
+    // Ensure required terms and conditions have been accepted, then attempt to commit
+    if (nullptr != termsAndConditionsProvider)
     {
-        SessionHandle handle = commandObj->GetExchangeContext()->GetSessionHandle();
-        // If not a CASE session, or the fabric does not match the fail-safe,
-        // error out.
-        if (handle->GetSessionType() != Session::SessionType::kSecure ||
-            handle->AsSecureSession()->GetSecureSessionType() != SecureSession::Type::kCASE ||
-            !failSafe.MatchesFabricIndex(commandObj->GetAccessingFabricIndex()))
+        response.errorCode = CheckTermsAndConditionsAcknowledgements(termsAndConditionsProvider);
```

Comment:
So that can return one of three values: NotMet, NotReceived, NotAccepted.  The spec allows NotReceived and NotMet here; it does not have any provisions for NotAccepted.

Is NotAccepted not a possible return value in this situation?  If so that should probably be documented here with an explanation of why not.

If it's a possible return value, this does not match the spec.

Reply:
Can you help explain again? The function can return all 4 states, unless I'm missing something.

1. kTCAcknowledgementsNotReceived if there was never a SetTCAcknowledgements command sent
2. kTCMinVersionNotMet if it received the SetTCAcknowledgements command,  but the version was too low
3. kRequiredTCNotAccepted if it received the SetTCAcknowledgements command,  but required bits were not accepted
4. kOk